crates/terminal/src/terminal.rs

```diff
@@ -1092,7 +1104,83 @@ impl Terminal {
         self.write_bytes_to_pty(input);
     }
 
+    pub fn toggle_vi_mode(&mut self) {
+        self.vi_mode = !self.vi_mode;
+        self.events.push_back(InternalEvent::ToggleViMode);
```

Comment:
I changed this in the patch I commented. I think it's better if we send the event and only toggle the boolean once we get the event back, otherwise we might have a data race.